File: core/src/main/java/vn/loitp/function/epub/core/EpubReaderReadActivity.java
package vn.loitp.function.epub.core;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.text.Html;
import android.util.Base64;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;

import loitp.core.R;
import vn.loitp.core.base.BaseFontActivity;
import vn.loitp.core.common.Constants;
import vn.loitp.core.utilities.LLog;
import vn.loitp.core.utilities.LReaderUtil;
import vn.loitp.core.utilities.LUIUtil;
import vn.loitp.function.epub.BookSection;
import vn.loitp.function.epub.CssStatus;
import vn.loitp.function.epub.Reader;
import vn.loitp.function.epub.exception.OutOfPagesException;
import vn.loitp.function.epub.exception.ReadingException;
import vn.loitp.function.epub.model.BookInfo;
import vn.loitp.views.LToast;
import vn.loitp.views.layout.floatdraglayout.DisplayUtil;
import vn.loitp.views.viewpager.viewpagertransformers.ZoomOutSlideTransformer;

public class EpubReaderReadActivity extends BaseFontActivity implements PageFragment.OnFragmentReadyListener {
    public static final String BOOK_INFO = "BOOK_INFO";
    public static final String IS_WEBVIEW = "IS_WEBVIEW";
    public static final String IS_USE_FONT = "IS_USE_FONT";
    private Reader reader;

    private ViewPager mViewPager;
    private SectionsPagerAdapter mSectionsPagerAdapter;
    private int pageCount = Integer.MAX_VALUE;
    private int pxScreenWidth;
    private boolean isPickedWebView = false;
    //private MenuItem searchMenuItem;
    //private SearchView searchView;
    private boolean isSkippedToPage = false;

    private BookInfo bookInfo;
    private boolean isUseFont = true;
    private TextView tvPage;
    private TextView tvTitle;
    private ImageView ivCover;
    private RelativeLayout rlSplash;
    private AdView adView;

    private void setCoverBitmap() {
        boolean isCoverImageNotExists = bookInfo.isCoverImageNotExists();
        if (!isCoverImageNotExists) {
            // Not searched for coverImage for this position yet. It may exist.
            Bitmap savedBitmap = bookInfo.getCoverImageBitmap();
            if (savedBitmap != null) {
                ivCover.setImageBitmap(savedBitmap);
            } else {
                byte[] coverImageAsBytes = bookInfo.getCoverImage();
                if (coverImageAsBytes != null) {
                    Bitmap bitmap = LReaderUtil.decodeBitmapFromByteArray(coverImageAsBytes, 100, 200);
                    bookInfo.setCoverImageBitmap(bitmap);
                    bookInfo.setCoverImage(null);
                    ivCover.setImageBitmap(bitmap);
                } else {
                    // Searched and not found.
                    bookInfo.setCoverImageNotExists(true);
                    ivCover.setImageResource(LReaderUtil.getDefaultCover());
                }
            }
        } else {
            // Searched before and not found.
            ivCover.setImageResource(LReaderUtil.getDefaultCover());
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        rlSplash = (RelativeLayout) findViewById(R.id.rl_splash);
        tvTitle = (TextView) findViewById(R.id.tv_title);
        ivCover = (ImageView) findViewById(R.id.iv_cover);
        LUIUtil.setTextShadow(tvTitle);
        bookInfo = (BookInfo) getIntent().getSerializableExtra(BOOK_INFO);
        if (bookInfo == null) {
            LToast.show(activity, getString(R.string.err_unknow));
            onBackPressed();
        }
        isPickedWebView = getIntent().getBooleanExtra(IS_WEBVIEW, true);
        isUseFont = getIntent().getBooleanExtra(IS_USE_FONT, true);
        setCoverBitmap();
        String titleBook = bookInfo.getTitle();
        if (titleBook == null) {
            titleBook = "Loading...";
        }
        tvTitle.setText(titleBook);
        pxScreenWidth = getResources().getDisplayMetrics().widthPixels;
        mSectionsPagerAdapter = new SectionsPagerAdapter(getSupportFragmentManager());
        mViewPager = (ViewPager) findViewById(R.id.container);
        LUIUtil.setPullLikeIOSHorizontal(mViewPager);
        tvPage = (TextView) findViewById(R.id.tv_page);
        mViewPager.setOffscreenPageLimit(0);
        mViewPager.setPageTransformer(true, new ZoomOutSlideTransformer());
        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            }

            @Override
            public void onPageSelected(int position) {
                tvPage.setText("" + position);
            }

            @Override
            public void onPageScrollStateChanged(int state) {
            }
        });
        final String adUnitId = getIntent().getStringExtra(Constants.AD_UNIT_ID_BANNER);
        LLog.d(TAG, "adUnitId " + adUnitId);
        LinearLayout lnAdview = (LinearLayout) findViewById(R.id.ln_adview);
        if (adUnitId == null || adUnitId.isEmpty()) {
            lnAdview.setVisibility(View.GONE);
        } else {
            adView = new AdView(activity);
            adView.setAdSize(AdSize.SMART_BANNER);
            adView.setAdUnitId(adUnitId);
            LUIUtil.createAdBanner(adView);
            lnAdview.addView(adView);
            int navigationHeight = DisplayUtil.getNavigationBarHeight(activity);
            LUIUtil.setMargins(lnAdview, 0, 0, 0, navigationHeight + navigationHeight / 3);
        }
        loadData = new LoadData();
        loadData.execute();
    }

    private LoadData loadData;

    private class LoadData extends AsyncTask<Void, Void, Void> {
        private int lastSavedPage = 0;

        @Override
        protected void onPreExecute() {
            //LLog.d(TAG, "onPreExecute");
            super.onPreExecute();
        }

        @Override
        protected Void doInBackground(Void... voids) {
            //LLog.d(TAG, "doInBackground");
            try {
                reader = new Reader();
                // Setting optionals once per file is enough.
                reader.setMaxContentPerSection(1250 * 10);
                reader.setCssStatus(isPickedWebView ? CssStatus.INCLUDE : CssStatus.OMIT);
                reader.setIsIncludingTextContent(true);
                reader.setIsOmittingTitleTag(true);
                // This method must be called before readSection.
                reader.setFullContent(bookInfo.getFilePath());
                // int lastSavedPage = reader.setFullContentWithProgress(filePath);
                if (reader.isSavedProgressFound()) {
                    lastSavedPage = reader.loadProgress();
                }

            } catch (ReadingException e) {
                LLog.e(TAG, "doInBackground " + e.toString());
                LToast.show(activity, "Error: " + e.getMessage());
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            //LLog.d(TAG, "onPostExecute");
            super.onPostExecute(aVoid);
            rlSplash.setVisibility(View.GONE);
            rlSplash = null;
            mViewPager.setAdapter(mSectionsPagerAdapter);
            if (reader.isSavedProgressFound()) {
                mViewPager.setCurrentItem(lastSavedPage);
            }
        }
    }

    @Override
    public void onPause() {
        if (adView != null) {
            adView.pause();
        }
        super.onPause();
    }

    @Override
    public void onResume() {
        if (adView != null) {
            adView.resume();
        }
        super.onResume();
    }

    @Override
    protected void onDestroy() {
        if (loadData == null) {
            loadData.cancel(true);
        }
        if (adView != null) {
            adView.destroy();
        }
        super.onDestroy();
    }

    @Override
    protected boolean setFullScreen() {
        return true;
    }

    @Override
    protected String setTag() {
        return "TAG" + getClass().getSimpleName();
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_epub_reader_read;
    }

    @Override
    public View onFragmentReady(int position) {
        BookSection bookSection = null;
        try {
            bookSection = reader.readSection(position);
        } catch (ReadingException e) {
            e.printStackTrace();
            LToast.show(activity, "Error: " + e.getMessage());
        } catch (OutOfPagesException e) {
            e.printStackTrace();
            this.pageCount = e.getPageCount();
            if (isSkippedToPage) {
                LToast.show(activity, "Max page number is: " + this.pageCount);
            }
            mSectionsPagerAdapter.notifyDataSetChanged();
        }
        isSkippedToPage = false;
        if (bookSection != null) {
            //LLog.d(TAG, "onFragmentReady: " + bookSection.getSectionContent());
            return setFragmentView(isPickedWebView, bookSection.getSectionContent(), "text/html", "UTF-8"); // reader.isContentStyled
        }
        return null;
    }

    /*@Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_epub_reader, menu);

        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        searchMenuItem = menu.findItem(R.id.action_search);
        searchView = (SearchView) MenuItemCompat.getActionView(searchMenuItem);
        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {

                if (query != null && !query.equals("")) {

                    if (TextUtils.isDigitsOnly(query)) {
                        loseFocusOnSearchView();

                        int skippingPage = Integer.valueOf(query);

                        if (skippingPage >= 0) {
                            isSkippedToPage = true;
                            mViewPager.setCurrentItem(skippingPage);
                        } else {
                            Toast.makeText(activity, "Page number can't be less than 0", Toast.LENGTH_LONG).show();
                        }

                    } else {
                        loseFocusOnSearchView();
                        Toast.makeText(activity, "Only numbers are allowed", Toast.LENGTH_LONG).show();
                    }

                    return true;
                }

                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });

        return true;
    }*/

    /*@Override
    public void onBackPressed() {
        if (!searchView.isIconified()) {
            loseFocusOnSearchView();
        } else {
            super.onBackPressed();
        }
    }*/

    @Override
    public void onStop() {
        super.onStop();
        try {
            reader.saveProgress(mViewPager.getCurrentItem());
            LToast.show(activity, "Saved page: " + mViewPager.getCurrentItem() + "...");
        } catch (ReadingException e) {
            e.printStackTrace();
            LToast.show(activity, "Progress is not saved: " + e.getMessage());
        } catch (OutOfPagesException e) {
            e.printStackTrace();
            LToast.show(activity, "Progress is not saved. Out of Bounds. Page Count: " + e.getPageCount());
        }
    }

    /*@Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_search) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }*/

    private String getStyledFont(String html) {
        if (!isUseFont) {
            return html;
        }
        boolean addBodyStart = !html.toLowerCase().contains("<body>");
        boolean addBodyEnd = !html.toLowerCase().contains("</body");
        return "<style type=\"text/css\">@font-face {font-family: CustomFont;" +
                "src: url(\"file:///android_asset/" +
                LUIUtil.getFontForAll() +
                "\")}" +
                "body {font-family: CustomFont;font-size: medium;text-align: justify;}</style>" +
                (addBodyStart ? "<body>" : "") + html + (addBodyEnd ? "</body>" : "");
    }

    private View setFragmentView(boolean isContentStyled, String data, String mimeType, String encoding) {
        FrameLayout.LayoutParams layoutParams = new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        if (isContentStyled) {
            WebView webView = new WebView(activity);
            //webView.loadDataWithBaseURL(null, data, mimeType, encoding, null);
            webView.loadDataWithBaseURL("file:///android_asset/",
                    getStyledFont(data),
                    mimeType, encoding, "about:blank");
//            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
//                webView.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
//            }
            webView.setLayoutParams(layoutParams);
            return webView;
        } else {
            ScrollView scrollView = new ScrollView(activity);
            LUIUtil.setPullLikeIOSVertical(scrollView);
            scrollView.setLayoutParams(layoutParams);
            TextView textView = new TextView(activity);
            textView.setTextColor(Color.BLACK);
            LUIUtil.setTextSize(textView, TypedValue.COMPLEX_UNIT_DIP, 18);
            textView.setLayoutParams(layoutParams);
            textView.setText(Html.fromHtml(data, new Html.ImageGetter() {
                @Override
                public Drawable getDrawable(String source) {
                    String imageAsStr = source.substring(source.indexOf(";base64,") + 8);
                    byte[] imageAsBytes = Base64.decode(imageAsStr, Base64.DEFAULT);
                    Bitmap imageAsBitmap = BitmapFactory.decodeByteArray(imageAsBytes, 0, imageAsBytes.length);

                    int imageWidthStartPx = (pxScreenWidth - imageAsBitmap.getWidth()) / 2;
                    int imageWidthEndPx = pxScreenWidth - imageWidthStartPx;

                    Drawable imageAsDrawable = new BitmapDrawable(getResources(), imageAsBitmap);
                    imageAsDrawable.setBounds(imageWidthStartPx, 0, imageWidthEndPx, imageAsBitmap.getHeight());
                    return imageAsDrawable;
                }
            }, null));
            int pxPadding = dpToPx(10);
            textView.setPadding(pxPadding, pxPadding, pxPadding, pxPadding);
            if (isUseFont) {
                Typeface face = Typeface.createFromAsset(getAssets(), LUIUtil.getFontForAll());
                textView.setTypeface(face);
            }
            scrollView.addView(textView);
            return scrollView;
        }
    }

    /*private void loseFocusOnSearchView() {
        searchView.setQuery("", false);
        searchView.clearFocus();
        searchView.setIconified(true);
        MenuItemCompat.collapseActionView(searchMenuItem);
    }*/

    private int dpToPx(int dp) {
        DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
        return Math.round(dp * (displayMetrics.xdpi / DisplayMetrics.DENSITY_DEFAULT));
    }

    public class SectionsPagerAdapter extends FragmentStatePagerAdapter {

        SectionsPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public int getCount() {
            return pageCount;
        }

        @Override
        public Fragment getItem(int position) {
            // getItem is called to instantiate the fragment for the given page.
            return PageFragment.newInstance(position);
        }
    }
}

File: core/src/main/java/com/function/epub/core/EpubReaderReadActivity.kt
package com.function.epub.core

import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Bundle
import android.text.Html
import android.text.Html.ImageGetter
import android.util.Base64
import android.view.View
import android.view.ViewGroup
import android.webkit.WebSettings
import android.widget.FrameLayout
import android.widget.ScrollView
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.ViewPager
import com.R
import com.annotation.IsFullScreen
import com.annotation.LogTag
import com.core.base.BaseApplication
import com.core.base.BaseFontActivity
import com.core.common.Constants
import com.core.utilities.*
import com.core.utilities.LReaderUtil.Companion.defaultCover
import com.daimajia.androidanimations.library.Techniques
import com.function.epub.BookSection
import com.function.epub.Reader
import com.function.epub.core.PageFragment.OnFragmentReadyListener
import com.function.epub.exception.OutOfPagesException
import com.function.epub.exception.ReadingException
import com.function.epub.model.BookInfo
import com.function.epub.model.BookInfoData
import com.function.epub.viewmodels.EpubViewModel
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.interfaces.CallbackAnimation
import com.utils.util.ConvertUtils
import com.views.LWebView
import com.views.setSafeOnClickListener
import com.views.viewpager.viewpagertransformers.ZoomOutSlideTransformer
import kotlinx.android.synthetic.main.l_activity_epub_reader_read.*

@LogTag("loitppEpubReaderReadActivity")
@IsFullScreen(false)
class EpubReaderReadActivity : BaseFontActivity(), OnFragmentReadyListener {

    companion object {
        private const val idWebView = 696969
    }

    private val reader = Reader()
    private var isSkippedToPage = false
    private var sectionsPagerAdapter: SectionsPagerAdapter? = null
    private var pageCount = Int.MAX_VALUE
    private val pxScreenWidth = LScreenUtil.screenWidth
    private var bookInfo: BookInfo? = null
    private var adView: AdView? = null
    private var epubViewModel: EpubViewModel? = null

    private val isDarkTheme = LSharedPrefsUtil.instance.getBoolean(Constants.KEY_IS_DARK_THEME, true)
//    private val isDarkTheme = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.l_activity_epub_reader_read)

        bookInfo = BookInfoData.instance.bookInfo
        if (bookInfo == null) {
            showShort(getString(R.string.err_unknow))
            onBackPressed()
        }

        setupViews()
        setupViewModels()
        setupTheme()

        bookInfo?.let {
            epubViewModel?.loadData(reader = reader, bookInfo = it)
        }
    }

    private fun setupViews() {
        setCoverBitmap()
        val titleBook = bookInfo?.title ?: getString(R.string.loading)
        tvTitle.text = titleBook

        sectionsPagerAdapter = SectionsPagerAdapter(supportFragmentManager)
        viewPager.apply {
            LUIUtil.setPullLikeIOSHorizontal(viewPager = this)
            this.offscreenPageLimit = 0
            this.setPageTransformer(true, ZoomOutSlideTransformer())
            this.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
                override fun onPageSelected(position: Int) {
                    tvPage.text = "$position"
                }

                override fun onPageScrollStateChanged(state: Int) {}
            })
            this.adapter = sectionsPagerAdapter
        }

        val adUnitId = intent.getStringExtra(Constants.AD_UNIT_ID_BANNER)
        if (adUnitId.isNullOrEmpty() || !LConnectivityUtil.isConnected()) {
            lnAdView.visibility = View.GONE
        } else {
            adView = AdView(this)
            adView?.let { av ->
                av.adSize = AdSize.SMART_BANNER
                av.adUnitId = adUnitId
                LUIUtil.createAdBanner(av)
                lnAdView.addView(av)
                lnAdView.requestLayout()
            }
        }
        btBack.setSafeOnClickListener {
            onBackPressed()
        }
        btZoomIn.setSafeOnClickListener {
            handleZoomIn()
        }
        btZoomOut.setSafeOnClickListener { view ->
            handleZoomOut()
        }
        llGuide.setSafeOnClickListener {
            handleGuide()
        }
    }

    private fun setupViewModels() {
        epubViewModel = getViewModel(EpubViewModel::class.java)
        epubViewModel?.let { vm ->
            vm.loadDataActionLiveData.observe(this, androidx.lifecycle.Observer { actionData ->
                logD("loadDataActionLiveData observe " + BaseApplication.gson.toJson(actionData))
                val isDoing = actionData.isDoing
                val isSuccess = actionData.isSuccess

                if (isDoing == false && isSuccess == true) {
                    LUIUtil.setDelay(mls = 1000, runnable = Runnable {
                        rlSplash?.visibility = View.GONE
                    })
                    sectionsPagerAdapter?.notifyDataSetChanged()
                    val lastSavedPage = actionData.data ?: 1
                    logD("loadDataActionLiveData lastSavedPage $lastSavedPage")
                    viewPager?.currentItem = lastSavedPage
                    if (lastSavedPage == 0) {
                        tvPage?.text = "0"
                    }
                    llGuide?.visibility = View.VISIBLE
                }
            })
        }
    }

    private fun setupTheme() {
        if (isDarkTheme) {
            layoutRootViewEpub.setBackgroundColor(Color.BLACK)
            layoutControl.setBackgroundColor(Color.BLACK)
            btBack.setColorFilter(Color.WHITE)
            tvPage.setTextColor(Color.WHITE)
            btZoomIn.setColorFilter(Color.WHITE)
            btZoomOut.setColorFilter(Color.WHITE)
            blurView.setOverlayColor(LAppResource.getColor(R.color.white35))
            tvTitle.setTextColor(Color.BLACK)
        } else {
            layoutRootViewEpub.setBackgroundColor(Color.WHITE)
            layoutControl.setBackgroundColor(Color.WHITE)
            btBack.setColorFilter(Color.BLACK)
            tvPage.setTextColor(Color.BLACK)
            btZoomIn.setColorFilter(Color.BLACK)
            btZoomOut.setColorFilter(Color.BLACK)
            blurView.setOverlayColor(LAppResource.getColor(R.color.black35))
            tvTitle.setTextColor(Color.WHITE)
        }
    }

    private fun handleZoomIn() {
        LAnimationUtil.play(view = btZoomIn, techniques = Techniques.Pulse)

        sectionsPagerAdapter?.let { adapter ->
            try {
                val pageFragment = adapter.instantiateItem(viewPager, viewPager.currentItem)
                if (pageFragment is PageFragment) {
                    zoomIn(pageFragment = pageFragment)
                }

                val pageFragmentNext = adapter.instantiateItem(viewPager, viewPager.currentItem + 1)
                if (pageFragmentNext is PageFragment) {
                    zoomIn(pageFragment = pageFragmentNext)
                }

                val pageFragmentPrev = adapter.instantiateItem(viewPager, viewPager.currentItem - 1)
                if (pageFragmentPrev is PageFragment) {
                    zoomIn(pageFragment = pageFragmentPrev)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun handleZoomOut() {
        LAnimationUtil.play(view = btZoomOut, techniques = Techniques.Pulse)
        sectionsPagerAdapter?.let { adapter ->
            try {
                val pageFragment = adapter.instantiateItem(viewPager, viewPager.currentItem)
                if (pageFragment is PageFragment) {
                    zoomOut(pageFragment)
                }

                val pageFragmentNext = adapter.instantiateItem(viewPager, viewPager.currentItem + 1)
                if (pageFragmentNext is PageFragment) {
                    zoomOut(pageFragmentNext)
                }

                val pageFragmentPrev = adapter.instantiateItem(viewPager, viewPager.currentItem - 1)
                if (pageFragmentPrev is PageFragment) {
                    zoomOut(pageFragmentPrev)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun handleGuide() {
        LAnimationUtil.play(view = llGuide, techniques = Techniques.SlideOutLeft, callbackAnimation = object : CallbackAnimation {
            override fun onCancel() {}
            override fun onEnd() {
                llGuide?.visibility = View.GONE
            }

            override fun onRepeat() {}
            override fun onStart() {}
        })
    }

    private fun setCoverBitmap() {
        bookInfo?.let { bi ->
            val isCoverImageNotExists = bi.isCoverImageNotExists
            if (isCoverImageNotExists) {
                // Searched before and not found.
                ivCover.setImageResource(defaultCover)
            } else {
                // Not searched for coverImage for this position yet. It may exist.
                val savedBitmap = bi.coverImageBitmap
                if (savedBitmap == null) {
                    val coverImageAsBytes = bi.coverImage
                    if (coverImageAsBytes == null) {
                        // Searched and not found.
                        bi.isCoverImageNotExists = true
                        ivCover.setImageResource(defaultCover)
                    } else {
                        val bitmap = LReaderUtil.decodeBitmapFromByteArray(coverImage = coverImageAsBytes, reqWidth = 100, reqHeight = 200)
                        bi.coverImageBitmap = bitmap
                        bi.coverImage = null
                        ivCover.setImageBitmap(bitmap)
                    }
                } else {
                    ivCover.setImageBitmap(savedBitmap)
                }
            }
        }
    }

    override fun onFragmentReady(position: Int): View? {
        logD("onFragmentReady position $position")
        var bookSection: BookSection? = null
        try {
            bookSection = reader.readSection(position)
        } catch (e: ReadingException) {
            logE("onFragmentReady ReadingException $e")
        } catch (e: OutOfPagesException) {
            logE("onFragmentReady OutOfPagesException $e")
            pageCount = e.pageCount
            if (isSkippedToPage) {
                showShort("Max page number is: $pageCount")
            }
            sectionsPagerAdapter?.notifyDataSetChanged()
        } catch (e: Exception) {
            logE("onFragmentReady Exception $e")
        }
        isSkippedToPage = false
        return if (bookSection == null) {
            null
        } else {
            setFragmentView(isContentStyled = true,
                    data = bookSection.sectionContent ?: "",
                    mimeType = "text/html",
                    encoding = "UTF-8")
        }
    }

    public override fun onPause() {
        adView?.pause()
        super.onPause()
    }

    public override fun onResume() {
        adView?.resume()
        super.onResume()
    }

    override fun onDestroy() {
        adView?.destroy()
        BookInfoData.instance.bookInfo = null
        super.onDestroy()
    }

    public override fun onStop() {
        super.onStop()
        try {
            reader.saveProgress(viewPager.currentItem)
            showShort(msg = "Saved page: " + viewPager.currentItem + "...")
        } catch (e: ReadingException) {
            e.printStackTrace()
            showShort(msg = "Progress is not saved: " + e.message)
        } catch (e: OutOfPagesException) {
            e.printStackTrace()
            showShort(msg = "Progress is not saved. Out of Bounds. Page Count: " + e.pageCount)
        }
    }

    inner class SectionsPagerAdapter internal constructor(fm: FragmentManager)
        : FragmentStatePagerAdapter(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

        override fun getCount(): Int {
            return pageCount
        }

        override fun getItem(position: Int): Fragment {
            return PageFragment.newInstance(position)
        }
    }

    @Suppress("DEPRECATION")
    private fun setFragmentView(isContentStyled: Boolean, data: String, mimeType: String, encoding: String): View {
        val layoutParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        logD("setFragmentView data $data")
        return if (isContentStyled) {
            val lWebView = LWebView(this)
            lWebView.apply {
                shouldOverrideUrlLoading(shouldOverrideUrlLoading = true)
                setEnableCopyContent(isEnableCopyContent = true)
                id = idWebView
//                logD(">>>setFragmentView data $data")
                val fontSizePx = LAppResource.getDimenValue(R.dimen.txt_small)
                val paddingPx = LAppResource.getDimenValue(R.dimen.padding_small)
//                logD(">>>setFragmentView fontSizePx $fontSizePx, paddingPx $paddingPx")
                if (isDarkTheme) {
                    loadDataString(bodyContent = data,
                            backgroundColor = "black",
                            textColor = "white",
                            textAlign = "justify",
                            fontSizePx = fontSizePx,
                            paddingPx = paddingPx
                    )
                } else {
                    loadDataString(bodyContent = data,
                            backgroundColor = "white",
                            textColor = "black",
                            textAlign = "justify",
                            fontSizePx = fontSizePx,
                            paddingPx = paddingPx
                    )
                }

                scrollBarSize = ConvertUtils.dp2px(2f)
                this.layoutParams = layoutParams
                lWebView.setTextSize(sizePercent = LPrefUtil.getTextSizePercentEpub())
                callback = object : LWebView.Callback {
                    override fun onScroll(l: Int, t: Int, oldl: Int, oldt: Int) {
                    }

                    override fun onScrollTopToBottom() {
//                    logD("onScrollTopToBottom")
                    }

                    override fun onScrollBottomToTop() {
//                    logD("onScrollBottomToTop")
                    }

                    override fun onProgressChanged(progress: Int) {
//                    logD("onProgressChanged $progress")
                    }

                    override fun shouldOverrideUrlLoading(url: String) {
//                    logD("shouldOverrideUrlLoading $url")
                    }

                }
            }
            lWebView
        } else {
            //this case wont occur
            val scrollView = ScrollView(this)
            scrollView.layoutParams = layoutParams
            val textView = TextView(this)
            textView.layoutParams = layoutParams
            textView.text = Html.fromHtml(data, ImageGetter { source ->
                val imageAsStr = source.substring(source.indexOf(";base64,") + 8)
                val imageAsBytes = Base64.decode(imageAsStr, Base64.DEFAULT)
                val imageAsBitmap = BitmapFactory.decodeByteArray(imageAsBytes, 0, imageAsBytes.size)
                val imageWidthStartPx = (pxScreenWidth - imageAsBitmap.width) / 2
                val imageWidthEndPx = pxScreenWidth - imageWidthStartPx
                val imageAsDrawable: Drawable = BitmapDrawable(resources, imageAsBitmap)
                imageAsDrawable.setBounds(imageWidthStartPx, 0, imageWidthEndPx, imageAsBitmap.height)
                imageAsDrawable
            }, null)
            val pxPadding = ConvertUtils.dp2px(12f)
            textView.setPadding(pxPadding, pxPadding, pxPadding, pxPadding)
            scrollView.addView(textView)
            scrollView
        }
    }

    @Suppress("DEPRECATION")
    private fun zoomIn(pageFragment: PageFragment?) {
        if (pageFragment == null || pageFragment.view == null) {
            return
        }
        val webView = pageFragment.view?.findViewById<LWebView>(idWebView)
        if (webView == null) {
            logE("webView null")
            return
        }
        val settings = webView.settings
        val currentApiVersion = Build.VERSION.SDK_INT
        if (currentApiVersion <= Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            settings.textSize = WebSettings.TextSize.LARGER
        } else {
            var size = (settings.textZoom * 1.1).toInt()
            if (size > 250) {
                size = 250
            }
            LPrefUtil.setTextSizePercentEpub(value = size)
            webView.setTextSize(sizePercent = size)
        }
    }

    @Suppress("DEPRECATION")
    private fun zoomOut(pageFragment: PageFragment?) {
        if (pageFragment == null || pageFragment.view == null) {
            return
        }
        val webView = pageFragment.view?.findViewById<LWebView>(idWebView) ?: return
        val settings = webView.settings
        val currentAiVersion = Build.VERSION.SDK_INT
        if (currentAiVersion <= Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            settings.textSize = WebSettings.TextSize.SMALLEST
        } else {
            var size = (settings.textZoom / 1.1).toInt()
            if (size < 50) {
                size = 50
            }
            LPrefUtil.setTextSizePercentEpub(value = size)
            webView.setTextSize(sizePercent = size)
        }
    }
}